src/app/zap-templates/templates/app/MatterIDL.zapt

```diff
@@ -61,40 +61,102 @@
     {{/unless}} {{asLowerCamelCase name~}}
     {{~#if isArray~}} [] {{~/if}} = {{code}};
   {{/chip_server_cluster_attributes}}
-  {{#chip_cluster_commands}}
-  {{#if arguments}}
+  {{#if (is_server side)}}
+  {{#all_incoming_commands_for_cluster name 'server'}}
+  {{#zcl_command_arguments}}
 
-  request struct {{asUpperCamelCase name}}Request {
-    {{#chip_cluster_command_arguments}}
+{{#first}}
+  request struct {{asUpperCamelCase parent.commandName}}Request {
+{{/first}}    
     {{> idl_structure_member}}
+{{#last}}
 
-    {{/chip_cluster_command_arguments}}
   }
+{{/last}}
+  {{/zcl_command_arguments}}
+  {{/all_incoming_commands_for_cluster}}
   {{/if}}
-  {{/chip_cluster_commands}}
-  {{#chip_cluster_responses}}
+  {{#if (is_client side)}}
+  {{#all_outgoing_commands_for_cluster name 'client'}}
+  {{#zcl_command_arguments}}
 
-  response struct {{asUpperCamelCase name}} {
-    {{#chip_cluster_response_arguments}}
+{{#first}}
+  request struct {{asUpperCamelCase parent.commandName}}Request {
+{{/first}}    
     {{> idl_structure_member}}
+{{#last}}
 
-    {{/chip_cluster_response_arguments}}
   }
-  {{/chip_cluster_responses}}
-  {{#chip_cluster_commands}}
+{{/last}}
+  {{/zcl_command_arguments}}
+  {{/all_outgoing_commands_for_cluster}}
+  {{/if}}
+{{#if (is_client side)}}
+  {{#all_incoming_commands_for_cluster name 'client'}}
+  {{#zcl_command_arguments}}
+
+{{#first}}
+  response struct {{asUpperCamelCase parent.commandName}} {
+{{/first}}    
+    {{> idl_structure_member}}
+{{#last}}
+
+  }
+{{/last}}
+  {{/zcl_command_arguments}}
+  {{/all_incoming_commands_for_cluster}}
+  {{/if}}
+  {{#if (is_server side)}}
+  {{#all_outgoing_commands_for_cluster name 'server'}}
+  {{#zcl_command_arguments}}
+
+{{#first}}
+  response struct {{asUpperCamelCase parent.commandName}} {
+{{/first}}    
+    {{> idl_structure_member}}
+{{#last}}
+
+  }
+{{/last}}
+  {{/zcl_command_arguments}}
+  {{/all_outgoing_commands_for_cluster}}
+  {{/if}}
+  {{#if (is_server side)}}
+  {{#all_incoming_commands_for_cluster name 'server'}}
+  {{#first}}
+
+  {{/first}}
+  {{#if mustUseTimedInvoke}}
+  timed command
+  {{~else}}
+  command
+  {{~/if}} {{asUpperCamelCase commandName}}(
+    {{~#zcl_command_arguments~}}
+    {{~#first~}}
+       {{asUpperCamelCase parent.commandName}}Request
+    {{~/first~}}
+    {{~/zcl_command_arguments~}}
+  ): {{#if responseName}}{{responseName}}{{else}}{{#if_command_has_response commandName ''}}{{commandName}}Response{{else}}{{#if_command_has_response commandName 'Request'}}{{replace_string commandName 'Request' ''}}Response{{else}}DefaultSuccess{{/if_command_has_response}}{{/if_command_has_response}}{{/if}} = {{code}};
```

Comment:
This is horrible, not least because yet it duplicates logic elsewhere for determining what the right response to a command is, and now they can get out of sync, and then the IDL will not match "reality".  We should have this determination in _one_ place.

Not a blocker for this PR, but please ensure an issue is filed to fix this.  Whether that's done by moving the heuristic into the XML loader, or by removing the heuristic and annotating all responses explicitly in the XML, the end result should be no heuristic on queries from the database and the database holding the right state.